### PR TITLE
Added OpenAI Pre-trained weights to CLIP Models

### PR DESCRIPTION
**Pull request by pbontrager**
Summary:
Added pre-trained flag to all CLIP models. This will pull pre-trained weights for any model there are open weights for, otherwise it'll provide a warning and ignore the flag. The flag pulls the OpenAI weights from S3 which are saved as state_dict and loads those weights. All of the weights have been tested to output the same outputs as their corresponding models in the OpenAI CLIP repo.

CLIP models that were configured differently from the OpenAI repo were updated to match; this was the ViT-l/14 model and all the ResNet models. Additionally, a parameter was added to the CLIP text encoder to be able to be configured to match the OpenAI configs.

Test plan:
I added new unit tests that call each CLIP model with a random image and tokens and check the first 5 features of the text and image embeddings against an expected output. Running `pytest tests/models/clip/test_clip.py -vv` will run the tests. I removed the cached models on the server before running the test to verify the pre-trained weights were correctly pulled from S3 during the test.
Example notebook to verify weights match: https://colab.research.google.com/drive/1ANLQ4tBP2ejuLWZNprLl_Z8iWqdBG_CX?usp=sharing

Fixes https://github.com/facebookresearch/multimodal/issues/384

### CODE REVIEW

File: torchmultimodal/models/clip/text_encoder.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ class CLIPTextEncoder(nn.Module):
         context_length (int): Maximum sequence length for Transforer.
         vocab_size (int): Vocab size.
         width (int): Embedding dimension for Transformer encoder.
+        dim_feedforward (int): Dimmension of the feedfoward networks.
```
Comment (RdoubleA) [inline torchmultimodal/models/clip/text_encoder.py:27]:
```suggestion
        dim_feedforward (int): Dimension of the feedfoward networks.
```

**Review by RdoubleA — APPROVED** [review]
Overall looks good to me. Are there any references for the updated ResNet parameters? Or for any the CLIP variations. Would be great to add some info on where they're from/where they're used in a docstring

**Comment by pbontrager** [comment]
> Overall looks good to me. Are there any references for the updated ResNet parameters? Or for any the CLIP variations. Would be great to add some info on where they're from/where they're used in a docstring

The updated parameters are reverse engineered from the saved checkpoints from the CLIP library. I'm not sure where the original values came from.

File: tests/models/clip/test_clip.py [hunk 2]
```diff
@@ -120,3 +131,201 @@ def test_clip_vit_forward(self, context_length):
...
+        )
+
+    def test_clip_vit_l14(self, data):
+        text, image, *_ = data
+        model = pretrained.clip_vit_l14(True)
+        model.eval()
+        with torch.no_grad():
+            actual_a_embedding, actual_b_embedding = model(image, text)
+
+        expected_a_embedding = torch.tensor([-0.0066, 0.0428, 0.0369, 0.0357, -0.0001])
```
Comment (ankitade) [inline tests/models/clip/test_clip.py:190]: why not just check for mean

File: tests/models/clip/test_clip.py [hunk 3]
```diff
@@ -120,3 +131,201 @@ def test_clip_vit_forward(self, context_length):
...
+        expected_b_embedding = torch.tensor(
+            [-0.0415, -0.0375, -0.0474, 0.0057, -0.0353]
+        )
+
+        assert_expected(
```
Comment (ankitade) [inline tests/models/clip/test_clip.py:195]: assert on size

File: torchmultimodal/models/clip/model.py [hunk 4]
```diff
@@ -88,85 +108,121 @@ def clip_vit_l14() -> CLIP:
...
 
 
-def clip_rn50() -> CLIP:
+def clip_rn50(pretrained: bool = False) -> CLIP:
     vision_encoder = ResNetForCLIP(
         layers=(3, 4, 6, 3),
         output_dim=1024,
-        heads=1024,
-        width=2048,
+        heads=32,
```
Comment (ankitade) [inline torchmultimodal/models/clip/model.py:124]: idk how were these values so off to begin with

File: torchmultimodal/models/clip/model.py [hunk 5]
```diff
@@ -88,85 +108,121 @@ def clip_vit_l14() -> CLIP:
...
-    return CLIP(vision_encoder, text_encoder)
+    clip = CLIP(vision_encoder, text_encoder)
+    if pretrained:
+        load_module_from_url(clip, CLIP_MODEL_MAPPING["rn50x64"])
+    return clip
 
 
 # Note: these models use torchvision's ResNet
-def clip_rn50_tv() -> CLIP:
+def clip_rn50_tv(pretrained: bool = False) -> CLIP:
```
Comment (ankitade) [inline torchmultimodal/models/clip/model.py:201]: remove the pretrained parameter to all tv ones, its confusing

File: torchmultimodal/models/clip/text_encoder.py [hunk 6]
```diff
@@ -24,6 +24,7 @@ class CLIPTextEncoder(nn.Module):
         context_length (int): Maximum sequence length for Transforer.
         vocab_size (int): Vocab size.
         width (int): Embedding dimension for Transformer encoder.
+        dim_feedforward (int): Dimension of the feedfoward networks.
```
Comment (ankitade) [inline torchmultimodal/models/clip/text_encoder.py:27]: did u check adding this isnt breaking anything internal in terms of args order

File: tests/models/clip/test_clip.py [hunk 7]
```diff
@@ -120,3 +131,201 @@ def test_clip_vit_forward(self, context_length):
...
+            actual=actual_b_embedding[0, :5],
+            expected=expected_b_embedding,
+            rtol=0,
+            atol=1e-4,
+        )
+
+    def test_clip_rn50x64(self, data):
+        text, *_, image = data
```
Comment (ankitade) [inline tests/models/clip/test_clip.py:310]: nit: can u call it same as the data name i.e. image_448 for easy reading

**Review by ankitade — COMMENTED** [review]
can u add a notebook showing equivalence of fwd pass with open ai clip models.

File: tests/models/clip/test_clip.py (hunk 2 above)
Comment (pbontrager) [reply tests/models/clip/test_clip.py:190]: Not sure which one is more unique, but that probably would have been a bit more compact.

File: tests/models/clip/test_clip.py (hunk 3 above)
Comment (pbontrager) [reply tests/models/clip/test_clip.py:195]: Of the embedding or of all the parameters? If it's of the embedding, it seems unlikely the embedding size could change without changing the values.

File: torchmultimodal/models/clip/model.py (hunk 5 above)
Comment (pbontrager) [reply torchmultimodal/models/clip/model.py:201]: I was trying to keep the "api" consistent for all the model functions. If someone had some kind of model factory code or config I could also see it being annoying that only some models have the pretrained flag.

File: torchmultimodal/models/clip/text_encoder.py (hunk 6 above)
Comment (pbontrager) [reply torchmultimodal/models/clip/text_encoder.py:27]: I did a search on the codebase for CLIPTextEncoder and didn't find anywhere referencing the module that used positional arguments.

File: tests/models/clip/test_clip.py (hunk 3 above)
Comment (ankitade) [reply tests/models/clip/test_clip.py:195]: > unlikely the embedding size could change without changing the values.

don't quite follow. if you are going to assert on the mean value, its good to add a check on the shape of the output embedding (bsz x projection dim)?

File: torchmultimodal/models/clip/model.py (hunk 5 above)
Comment (ankitade) [reply torchmultimodal/models/clip/model.py:201]: in this case, I would say, functionality is more important than consistency. why would we expose an arg that does nothing (and gives an impression its doing something if the user misses the warning log)

**Review by ankitade — APPROVED** [review]
